File: leetcode-solutions/ruby/383. Ransom Note.rb
# @param {String} ransom_note
# @param {String} magazine
# @return {Boolean}
def can_construct(ransom_note, magazine)
    hash = ransom_note.chars.each_with_object(Hash.new(0)){|c, h| h[c] += 1}
    
    magazine.chars.each do |c|
        hash[c] -= 1
    end
    
    hash.values().select{|n| n > 0}.size == 0
end

File: leetcode-solutions/python/383. Ransom Note.py
# https://leetcode.com/problems/ransom-note/

class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        if len(ransomNote) > len(magazine):
            return False
        
        memo = {}
        for i, c in enumerate(magazine):
            if c in memo:
                memo[c] += 1
            else:
                memo[c] = 1

        for i, c in enumerate(ransomNote):
            if c not in memo or memo[c] == 0:
                return False
            memo[c] -= 1
        return True
                